loki/transformations/single_column/vertical.py

```diff
@@ -0,0 +1,218 @@
+        # loop interchange to expose vertical loops as outermost loops
+        loop_interchange(routine)
+        # handle initialization of arrays "jk +/- 1" arrays
+        multilevel_relevant_local_arrays_names = set(arr.name.lower() for arr in multilevel_relevant_local_arrays)
+        self.correct_init_of_multilevel_arrays(routine, multilevel_relevant_local_arrays_names)
+        # fuse vertical loops
+        loop_fusion(routine)
+        # demote in vertical dimension if possible
+        relevant_local_arrays_names = set(arr.name.lower() for arr in relevant_local_arrays)
+        demote_candidates = relevant_local_arrays_names - multilevel_relevant_local_arrays_names
+        # check which variables are safe to demote in the vertical
+        safe_to_demote = self.check_safe_to_demote(routine, demote_candidates)
+        # demote locals in vertical dimension
+        dimensions_to_demote = self.vertical.size_expressions + (f"{self.vertical.size}+1",)
+        demote_variables(routine, safe_to_demote, dimensions_to_demote)
+
+    def check_safe_to_demote(self, routine, demote_candidates):
+        """
+        Check whether variables that are candidates to be demoted in the vertical dimension are really
+        safe to be demoted.
+
+        Current heuristic: If the candidate is used in more than one vertical loop, assume it is NOT safe
+        to demote!
+        """
+        loop_var_map = {}
+        with pragmas_attached(routine, ir.Loop):
+            for loop in FindNodes(ir.Loop).visit(routine.body):
+                if loop.variable.name.lower() == self.vertical.index.lower():
+                    ignore = False
+                    if is_loki_pragma(loop.pragma, starts_with='fused-loop'):
+                        parameters = get_pragma_parameters(loop.pragma, starts_with='fused-loop')
+                        group = parameters.get('group', 'default')
+                        if group == 'ignore':
+                            ignore = True
+                    if not ignore:
+                        loop_var_map[loop] = as_tuple(var.name.lower() for var in FindVariables().visit(loop.body)
+                                if isinstance(var, sym.Array))
+
+        safe_to_demote = ()
+        for var in demote_candidates:
+            count = 0
+            for _, var_names in loop_var_map.items():
+                if var in var_names:
+                    count += 1
+            if count <= 1:
+                safe_to_demote += (var,)
+
+        return safe_to_demote
+
+    def find_relevant_local_arrays(self, routine):
+        """
+        Find local arrays/temporaries that do have the vertical dimension.
+        """
+        # local/temporary arrays
+        arrays = [var for var in FindVariables(unique=False).visit(routine.body) if isinstance(var, sym.Array)]
+        argument_names = [arg.name.lower() for arg in routine.arguments]
+        local_arrays = [arr for arr in arrays if arr.name.lower() not in argument_names]
```

Comment:
There's a common utility for this now: https://github.com/ecmwf-ifs/loki/blob/main/loki/transformations/utilities.py#L628

Could this be used here?